src/base/AsyncCallbacks.h

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (C) 1996-2022 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_BASE_ASYNCCALLBACKS_H
+#define SQUID_SRC_BASE_ASYNCCALLBACKS_H
+
+#include "base/AsyncCall.h"
+#include "base/AsyncJobCalls.h"
+#include "base/TypeTraits.h"
+
+/// access to a callback result carried by an asynchronous CallDialer
+template <typename AnswerT>
+class WithAnswer
+{
+public:
+    using Answer = AnswerT;
+
+    virtual ~WithAnswer() = default;
+
+    /// callback results setter
+    virtual Answer &answer() = 0;
+};
+
+/// a smart AsyncCall pointer for delivery of future results
+template <typename Answer>
+class AsyncCallback
+{
+public:
+    // all generated copying/moving functions are correct
+    AsyncCallback() = default;
+
+    template <class Call>
+    explicit AsyncCallback(const RefCount<Call> &call):
+        call_(call),
+        answer_(&(*call).dialer.answer())
```

Comment:
It is unclear why we need to store `answer_` separately, which can be obtained from already stored `call_`.

Also a minor nitpicking: I would  rewrite this line a little as `&(call->dialer.answer())`, to make it clear that we return answer's address.

Reply:
> It is unclear why we need to store `answer_` separately, which can be obtained from already stored `call_`.

The answer cannot be obtained (safely and fast). The type of `call_` is AsyncCall::Pointer. There is no AsyncCall::dialer (because generic AsyncCall does not know the type of its actual dialer).
 
> Also a minor nitpicking: I would rewrite this line a little as `&(call->dialer.answer())`, to make it clear that we return answer's address.

I agree, thank you. Trying that in 98d4a75dd94ca53e87185a3188661530fba46db3.